### ISSUE TITLE
x/tools/crypto/ssh: handshake: deadlock if other side drops/resets connection during rekeying

### ISSUE THREAD

**Issue by a-palchikov** [golang/go#11882]
The handshakeTransport.readLoop can exit with t.cond locked and any handshakeTransport.writePacket calls will block indefinitely at t.cond.Wait().

Additionally, there's a possible deadlock in handshakeTransport.writePacket that leaves t.mu locked:

    if t.writeError != nil {
        // t.mu.Unlock()  // <- missing
        return t.writeError
    }

**Comment by ianlancetaylor** [comment]
CC @agl @hanwen@google.com

**Comment by hanwen** [comment]
@ianlancetaylor - I am 'hanwen' on github, but for some reason the @ mention doesn't work. Do I have to be added to some list?

**Comment by ianlancetaylor** [comment]
@hanwen I can't claim to understand github, but I sent you an invitation that may do the trick.

**Comment by a-palchikov** [comment]
Sorry to bug you but I should've initially created two issues. The deadlock with handshake.readLoop exiting due to an error during rekey attempt and handshake.writePacket being blocked on t.cond is still imminent. Do I need to create another issue for this?

**Comment by hanwen** [comment]
can you make a separate issue?
